### PR TITLE
Not preserve the ownership of the files copied in the autoinstrumentation

### LINKED ISSUE #2655: Node.js pod CrashLoopBackOff after auto-instrumenting
### Component(s)

_No response_

## Description

Node.js application enters CrashLoop backoff when auto-instrumentation is enabled:

```yaml
apiVersion: v1
kind: Pod
metadata:
  annotations:
    instrumentation.opentelemetry.io/container-names: unleash
    instrumentation.opentelemetry.io/inject-nodejs: my-system/management-features
  creationTimestamp: "2024-02-21T19:48:21Z"
  generateName: my-demo-b64c6d87b-
  labels:
    app.kubernetes.io/created-by: controller-manager
    app.kubernetes.io/instance: my-demo
    app.kubernetes.io/name: Unleash
    app.kubernetes.io/part-of: unleasherator
    pod-template-hash: b64c6d87b
  name: my-demo-b64c6d87b-w72hh
  namespace: bifrost-unleash
  ownerReferences:
  - apiVersion: apps/v1
    blockOwnerDeletion: true
    controller: true
    kind: ReplicaSet
    name: my-demo-b64c6d87b
    uid: bd441ad4-6ada-4e92-b2fd-ec17a9ae9a44
  resourceVersion: "520000502"
  uid: 4a93c92e-3df2-4fc3-a611-1bdd4abe1ee6
spec:
  containers:
  - env:
    - name: INIT_ADMIN_API_TOKENS
      valueFrom:
        secretKeyRef:
          key: token
          name: unleasherator-my-demo-admin-key
    - name: DATABASE_PASS
      valueFrom:
        secretKeyRef:
          key: POSTGRES_PASSWORD
          name: my-demo
    - name: DATABASE_USER
      valueFrom:
        secretKeyRef:
          key: POSTGRES_USER
          name: my-demo
    - name: DATABASE_NAME
      valueFrom:
        secretKeyRef:
          key: POSTGRES_DB
          name: my-demo
    - name: DATABASE_HOST
      value: localhost
    - name: DATABASE_PORT
      value: "5432"
    - name: DATABASE_SSL
      value: "false"
    - name: DATABASE_URL
      value: postgres://$(DATABASE_USER):$(DATABASE_PASS)@$(DATABASE_HOST):$(DATABASE_PORT)/$(DATABASE_NAME)
    - name: GOOGLE_IAP_AUDIENCE
      value: /projects/898056957967/global/backendServices/6771496285844745965
    - name: TEAMS_API_URL
      value: http://teams-backend.my-system.svc/query
    - name: TEAMS_API_TOKEN
      valueFrom:
        secretKeyRef:
          key: token
          name: teams-api-token
    - name: TEAMS_ALLOWED_TEAMS
      value: aura,frontendplattform
    - name: LOG_LEVEL
      value: warn
    - name: DATABASE_POOL_MAX
      value: "3"
    - name: DATABASE_POOL_IDLE_TIMEOUT_MS
      value: "1000"
    - name: NODE_OPTIONS
      value: ' --require /otel-auto-instrumentation-nodejs/autoinstrumentation.js'
    - name: OTEL_SERVICE_NAME
      value: my-demo
    - name: OTEL_EXPORTER_OTLP_ENDPOINT
      value: http://opentelemetry-management-collector.my-system:4317
    - name: OTEL_RESOURCE_ATTRIBUTES_POD_NAME
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: metadata.name
    - name: OTEL_RESOURCE_ATTRIBUTES_NODE_NAME
      valueFrom:
        fieldRef:
          apiVersion: v1
          fieldPath: spec.nodeName
    - name: OTEL_PROPAGATORS
      value: tracecontext,baggage,b3
    - name: OTEL_RESOURCE_ATTRIBUTES
      value: k8s.container.name=unleash,k8s.deployment.name=my-demo,k8s.namespace.name=bifrost-unleash,k8s.node.name=$(OTEL_RESOURCE_ATTRIBUTES_NODE_NAME),k8s.pod.name=$(OTEL_RESOURCE_ATTRIBUTES_POD_NAME),k8s.replicaset.name=my-demo-b64c6d87b,service.version=v5.8.2-20240130-115753-fd5cd41
    image: europe-north1-docker.pkg.dev/my-io/my/images/unleash-v4:v5.8.2-20240130-115753-fd5cd41
    imagePullPolicy: Always
    livenessProbe:
      failureThreshold: 3
      httpGet:
        path: /health
        port: 4242
        scheme: HTTP
      initialDelaySeconds: 5
      periodSeconds: 10
      successThreshold: 1
      timeoutSeconds: 10
    name: unleash
    ports:
    - containerPort: 4242
      name: http
      protocol: TCP
    resources:
      limits:
        memory: 256Mi
      requests:
        cpu: 100m
        memory: 128Mi
    securityContext:
      allowPrivilegeEscalation: false
      capabilities:
        drop:
        - ALL
      runAsNonRoot: true
      runAsUser: 1001
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-xfgkk
      readOnly: true
    - mountPath: /otel-auto-instrumentation-nodejs
      name: opentelemetry-auto-instrumentation-nodejs
  - args:
    - --structured-logs
    - --port=5432
    - my-management-233d:europe-north1:bifrost-3de70742
    image: gcr.io/cloud-sql-connectors/cloud-sql-proxy:2.1.0
    imagePullPolicy: IfNotPresent
    name: sql-proxy
    resources:
      limits:
        memory: 100Mi
      requests:
        cpu: 10m
        memory: 100Mi
    securityContext:
      allowPrivilegeEscalation: false
      capabilities:
        drop:
        - ALL
      privileged: false
      runAsNonRoot: true
      runAsUser: 65532
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-xfgkk
      readOnly: true
  dnsPolicy: ClusterFirst
  enableServiceLinks: true
  initContainers:
  - command:
    - cp
    - -a
    - /autoinstrumentation/.
    - /otel-auto-instrumentation-nodejs
    image: ghcr.io/open-telemetry/opentelemetry-operator/autoinstrumentation-nodejs:0.46.0
    imagePullPolicy: IfNotPresent
    name: opentelemetry-auto-instrumentation-nodejs
    resources:
      limits:
        cpu: 500m
        memory: 128Mi
      requests:
        cpu: 50m
        memory: 128Mi
    securityContext:
      allowPrivilegeEscalation: false
      capabilities:
        drop:
        - ALL
      runAsNonRoot: true
      runAsUser: 1001
    terminationMessagePath: /dev/termination-log
    terminationMessagePolicy: File
    volumeMounts:
    - mountPath: /otel-auto-instrumentation-nodejs
      name: opentelemetry-auto-instrumentation-nodejs
    - mountPath: /var/run/secrets/kubernetes.io/serviceaccount
      name: kube-api-access-xfgkk
      readOnly: true
  nodeName: gke-my-management--nap-e2-standard--7ff15a1a-9bqj
  preemptionPolicy: PreemptLowerPriority
  priority: 0
  restartPolicy: Always
  schedulerName: gke.io/optimize-utilization-scheduler
  securityContext:
    runAsNonRoot: true
    seccompProfile:
      type: RuntimeDefault
  serviceAccount: bifrost-unleash-sql-user
  serviceAccountName: bifrost-unleash-sql-user
  terminationGracePeriodSeconds: 30
  tolerations:
  - effect: NoExecute
    key: node.kubernetes.io/not-ready
    operator: Exists
    tolerationSeconds: 300
  - effect: NoExecute
    key: node.kubernetes.io/unreachable
    operator: Exists
    tolerationSeconds: 300
  volumes:
  - name: kube-api-access-xfgkk
    projected:
      defaultMode: 420
      sources:
      - serviceAccountToken:
          expirationSeconds: 3607
          path: token
      - configMap:
          items:
          - key: ca.crt
            path: ca.crt
          name: kube-root-ca.crt
      - downwardAPI:
          items:
          - fieldRef:
              apiVersion: v1
              fieldPath: metadata.namespace
            path: namespace
  - emptyDir:
      sizeLimit: 200Mi
    name: opentelemetry-auto-instrumentation-nodejs
```

## Steps to Reproduce

Enable auto-instrumentation of a nodejs application like the one above.

## Expected Result

It should not crash.

## Actual Result

It fails to start with the following error:

```
cp: can't preserve ownership of '...': Operation not permitted
```

### Kubernetes Version

v1.28.3

### Operator version

0.93.0

### Collector version

latest

## Environment
Cloud: GKE

### Log output

```shell
cp: can't preserve ownership of '/otel-auto-instrumentation-nodejs/./autoinstrumentation.js': Operation not permitted
cp: can't preserve ownership of '/otel-auto-instrumentation-nodejs/./autoinstrumentation.d.ts.map': Operation not permitted
...
```

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by iblancasa**
Closes #2655.

Using `-a` is a problem when you try to instrument pods that are not run as root.

### CODE REVIEW

File: .chloggen/fix-2655.yaml [hunk 1]
```diff
@@ -0,0 +1,16 @@
+# One of 'breaking', 'deprecation', 'new_component', 'enhancement', 'bug_fix'
+change_type: bug_fix
+
+# The name of the component, or a single word describing the area of concern, (e.g. operator, target allocator, github action)
+component: instrumentation
+
+# A brief description of the change. Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: "Not preserve the ownership of the files from the autoinstrumenation when copied to the volume. This avoids isues when instrumenting workloads runned as non-root"
```
Comment (swiatekm-sumo) [inline .chloggen/fix-2655.yaml:8]:
```suggestion
note: "Don't preserve ownership of files copied from the autoinstrumenation image. This avoids issues when instrumenting workloads running as non-root"
```

File: pkg/instrumentation/dotnet.go [hunk 2]
```diff
@@ -127,7 +127,7 @@ func injectDotNetSDK(dotNetSpec v1alpha1.DotNet, pod corev1.Pod, index int, runt
 		pod.Spec.InitContainers = append(pod.Spec.InitContainers, corev1.Container{
 			Name:      dotnetInitContainerName,
 			Image:     dotNetSpec.Image,
-			Command:   []string{"cp", "-a", "/autoinstrumentation/.", dotnetInstrMountPath},
+			Command:   []string{"cp", "-r", "/autoinstrumentation/.", dotnetInstrMountPath},
```
Comment (swiatekm-sumo) [inline pkg/instrumentation/dotnet.go:130]: What *is* the ownership of these files after the change? Before, it was the owner of the file in the image, which depends on how it's built. Now, it's going to be the user we run the initcontainer as. Is that well defined?

Comment (iblancasa) [reply pkg/instrumentation/dotnet.go:130]: Yes. When I was fixing the instrumentation for other languages, I found the same problem and we fixed it just copying the folder without preserving the ownership -it is actually not needed-.

**Comment by swiatekm-sumo** [comment]
> @iblancasa would it be possible to build e2e tests to verify this?

We should consider running instrumentation e2e tests with images built from the repo imo at least in some circumstances. Not clear how annoying it would be.

**Comment by iblancasa** [comment]
> @iblancasa would it be possible to build e2e tests to verify this?

I'm not sure what to add to the current E2E tests to check this. Any ideas?

**Comment by swiatekm-sumo** [comment]
> > @iblancasa would it be possible to build e2e tests to verify this?
> 
> I'm not sure what to add to the current E2E tests to check this. Any ideas?

I don't think there's an easy way to do it right now. Have you tried building these images locally and running E2E tests with them?

**Comment by iblancasa** [comment]
> > > @iblancasa would it be possible to build e2e tests to verify this?
> > 
> > 
> > I'm not sure what to add to the current E2E tests to check this. Any ideas?
> 
> I don't think there's an easy way to do it right now. Have you tried building these images locally and running E2E tests with them?

This approach should be OK with the current images, right? We're not changing them.
I can test if, after this change, the applications from the E2E tests are still properly autoinstrumented. What do you think?

**Comment by swiatekm-sumo** [comment]
I thought for some reason we were changing the images themselves in this PR, sorry for the confusion. Existing E2E tests should probably be fine here, though I'm not 100% confident we have the necessary coverage for all instrumentations.

**Comment by iblancasa** [comment]
> ...though I'm not 100% confident we have the necessary coverage for all instrumentations.

What do you think I can add to increase the coverage?

**Comment by IshwarKanse** [comment]
@swiatekm-sumo @iblancasa  This https://github.com/open-telemetry/opentelemetry-operator/pull/2702  PR fixes our instrumentation tests and checks if all the containers in the pods are started and not in Error or CrashLoopBackoff state. If any of the containers in the instrumented application pod fails, we collect all the container logs using catch:

**Review by swiatekm-sumo — APPROVED** [review]
LGTM. E2E tests actually do check this, so if they pass, we should be good.

**Comment by iblancasa** [comment]
Can we merge this?

File: pkg/instrumentation/sdk_test.go [hunk 3]
```diff
@@ -865,7 +865,7 @@ func TestInjectDotNet(t *testing.T) {
 				{
 					Name:    dotnetInitContainerName,
 					Image:   "img:1",
-					Command: []string{"cp", "-a", "/autoinstrumentation/.", dotnetInstrMountPath},
+					Command: []string{"cp", "-r", "/autoinstrumentation/.", dotnetInstrMountPath},
```
Comment (jaronoff97) [inline pkg/instrumentation/sdk_test.go:868]: is cp -r available on the distributions we use? I tested it on my mac and it failed

Comment (swiatekm-sumo) [reply pkg/instrumentation/sdk_test.go:868]: I think we use busybox everywhere, so it should work.

Comment (jaronoff97) [reply pkg/instrumentation/sdk_test.go:868]: yeah i think just some extra confirmation that this worked on a local kind cluster would be good

Comment (iblancasa) [reply pkg/instrumentation/sdk_test.go:868]: The current E2E are testing this if I'm not wrong @IshwarKanse

Comment (IshwarKanse) [reply pkg/instrumentation/sdk_test.go:868]: So the current e2e tests are not checking the status of the containers in the pod. I have fixed that with the PR https://github.com/open-telemetry/opentelemetry-operator/pull/2702 with the changes, we check all the containers in the instrumented app pod are running and the test will fail if any of the container fails.

**Comment by sanicheev** [comment]
@iblancasa having the same issues.
Tested this branch on one of the clusters and it works.
Can we please merge this one?

**Comment by iblancasa** [comment]
> @iblancasa having the same issues. Tested this branch on one of the clusters and it works. Can we please merge this one?

Happy to heard it works!

**Pull request merged**